A30-xds-v3.md

```diff
@@ -0,0 +1,255 @@
+A30: xDS v3 Support
+----
+* Author(s): Mark D. Roth (markdroth)
+* Approver: ejona86, dfawley
+* Status: In Review
+* Implemented in: 
+* Last updated: 2020-06-25
+* Discussion at: https://groups.google.com/g/grpc-io/c/YRVqRMUQwhM
+
+## Abstract
+
+The purpose of this proposal is to describe how gRPC will handle the
+transition from v2 to v3 of the xDS API and to document how gRPC will handle
+xDS versioning on an ongoing basis.
+
+## Background
+
+gRPC has recently added support for obtaining configuration via the
+[xDS API](https://github.com/envoyproxy/data-plane-api).  Our current
+implementation uses v2 of the xDS API, but we are now adding support
+for [xDS
+v3](https://docs.google.com/document/d/1xeVvJ6KjFBkNjVspPbY_PwEDHC7XPi0J5p1SqUXcCl8/edit).
+In addition, we have been working with the xDS community on the new
+[minor/patch versioning
+proposal](https://docs.google.com/document/d/1afQ9wthJxofwOMCAle8qF-ckEfNWUuWtxbmdhxji0sI/edit),
+which should provide a smoother transition mechanism going forward.
+
+Regardless of whether a new version is indicated by bumping the
+major version (as was done with the v2 to v3 transition) or by bumping
+the minor version (as will probably be done going forward), the current
+plan is that there will still be one new version of the xDS API per
+year.
+
+### Related Proposals: 
+
+[A27: xDS-Based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+
+## Proposal
+
+### Version Support Policy
+
+gRPC's policy will be to support two versions of xDS at any given time.
+So if gRPC currently supports versions N-1 and N, then when xDS version
+N+1 is released, gRPC will add support for version N+1 and drop support
+for version N-1 at the same time, so that we will support versions N
+and N+1.
+
+This policy will give users 1-2 years to upgrade their xDS servers in
+order to continue using the latest release of gRPC.  Older releases of
+gRPC will of course continue to function using the older versions of the
+xDS API.
+
+### Deciding What Version to Use
+
+Any given gRPC release will support two versions of xDS, versions N-1 and N.
+It will assume that all servers speak version N-1, and there will be a
+server feature defined in the bootstrap file (see below) to tell it that
+the server supports version N.  When speaking to the server, gRPC will use
+the highest version that is common to both itself and the server.
```

Comment:
The minor versioning proposal features version negotiation, should this be the preferred mechanism all things being equal?

Reply:
The minor versioning proposal says that there will be some mechanism for version negotiation, but it does not actually define what that mechanism will be.  The current text of the proposal suggests using context params, but in offline conversation, you and I and Matt agreed that that won't actually work due to the exact-match semantics of context_params.  I believe that where we left this was that when someone actually gets around to implementing the minor version proposal, they will need to come up with a concrete proposal for version negotiation.  So I think that how this works is an open question, and I don't want to say anything about that here until we have a concrete proposal to evaluate.

It is also worth noting that the way our implementation is going to work, the version configured in the bootstrap file is effectively the transport version, not the resource version, whereas the version negotiation mechanism is by necessity primarily about resource version.  As an example, the version negotiation mechanism probably won't be usable to control things like the "xds transport next steps" design that you're currently working on implementing, because the client needs to know whether to use it before it sends the initial request containing its supported versions.  This is another reason why I believe it would be beneficial to consider the transport version and the resource version independently: in terms of mechanism, transport changes are generally going to be configured via client-side configuration rather than version negotiation.